### PR TITLE
feat(scheduler): base target methods and lambda invoke target

### PR DESCRIPTION
**Pull request by filletofish**
This PR contains implementation for Schedule Targets:

1. Creates a separate module for targets
2. Support imported resources, but not cross account, cross region resources as we discussed in RFC. The unit tests should cover 4 cases (target and role within the same stack, target is imported, role is imported, target and role are imported), 
3. I have moved out class `Schedule` from private package to depend on it in `schedule-targets` unit tests. 

Implementation is based on RFC: https://github.com/aws/aws-cdk-rfcs/blob/master/text/0474-event-bridge-scheduler-l2.md

Advances https://github.com/aws/aws-cdk/issues/23394

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-scheduler-alpha/test/group.test.ts [hunk 1]
```diff
@@ -101,18 +102,18 @@ describe('Schedule Group', () => {
     const schedule1 = new Schedule(stack, 'MyScheduleDummy1', {
       schedule: expr,
       group: group,
-      target: new targets.LambdaInvoke({
+      target: new targets.LambdaInvoke(func, {
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/test/group.test.ts:105]: Putting target resource before properties. I think it makes more sense than other way around.

File: packages/@aws-cdk/aws-scheduler-alpha/test/target.test.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+describe('schedule target', () => {
+  let stack: Stack;
+  let func: lambda.IFunction;
+  const expr = ScheduleExpression.at(new Date(Date.UTC(1969, 10, 20, 0, 0, 0)));
+
+  beforeEach(() => {
+    const app = new App();
+    stack = new Stack(app, 'Stack', { env: { region: 'us-east-1', account: '123456789012' } });
+    func = lambda.Function.fromFunctionArn(stack, 'Function', 'arn:aws:lambda:us-east-1:123456789012:function/somefunc');
+  });
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/test/target.test.ts:17]: Created role and stack in the same account. Otherwise `lambda.Function.grantInvoke` fails as it assumes that function is imported (from a different account).

Also, `grantInvoke` method creates permission on Lambda function itself (updates resource policy) if the role is in the different account.

File: packages/@aws-cdk/aws-scheduler-alpha/test/group.test.ts (hunk 1 above)
Comment (Jacco) [reply packages/@aws-cdk/aws-scheduler-alpha/test/group.test.ts:105]: Agreed :-)

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+import { Construct, IConstruct } from 'constructs';
+import { ScheduleTargetInput } from './input';
+import { ISchedule } from './schedule';
+
+export interface IScheduleTarget {
+  bind(_schedule: ISchedule): ScheduleTargetConfig;
+}
+
+export interface ScheduleTargetBaseProps {
+  readonly role?: iam.IRole;
```
Comment (Jacco) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:15]: All properties need some documentation

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+   *
+   * @default 185
+   */
+  readonly retryAttempts?: number;
+}
+
+/**
+ * Properties for an Schedule Target
+ */
+export interface ScheduleTargetConfig {
```
Comment (Jacco) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:50]: While you removal makes the code a bit easier, the use when specifying paramaters for a specific target becomes a bit harder because codecompletion will show you a lot of options to choose from while only one of the parameter properties should be filled in any specific case.

I personally don't like it. I think it is fine for CDK internal code to be complicated if it enhanced the use by users of the library.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+/**
+ * Obtain the Role for the EventBridge Schedule
+ *
+ * If a role already exists, it will be returned. This ensures that if multiple
+ * events have the same target, they will share a role.
+ * @internal
+ */
+export function singletonEventRole(scope: IConstruct): iam.IRole {
+  const id = 'SchedulesRole';
```
Comment (Jacco) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:121]: In the code we prepared this had a lot more beef.

1. You use one role named 'SchedulesRole' for all targets in the current stack. The more targets there are the less "least privilege' this will get. Personally I think one role per target makes more sense then one role per stack.
2. The code we started with had Stack.of(schedule) and then stack.node.tryFindChild(id) in stead of scope.node.tryFindChild. If you find the role for a target it is fine.

Comment (Jacco) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:121]: 3. Why do we not also narrow down of the use of the role to "aws:SourceArn" "aws:SourceAccount"

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts (hunk 4 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:50]: Did you maybe confuse it  with `ScheduleTargetBaseProps`? The interface that the customers will use is `ScheduleTargetBaseProps`, while the interface in discussion - `ScheduleTargetConfig` is used as output of `bind` operation. I think the documentation is misleading and I should update it.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts (hunk 5 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:121]: I originally preferred this implementation inspired by `events.target` (https://github.com/aws/aws-cdk/blob/e479bd4353aefa5e48189d2c71f6067489afe141/packages/aws-cdk-lib/aws-events-targets/lib/util.ts#L63-L81) as it seemed simpler to me. However, I agree that the least privilege policy is a good point. Now I understand where the complexity is coming from.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts (hunk 4 above)
Comment (Jacco) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:50]: I see you point now. I guess I just have to play around with it a bit :-)

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+function singletonScheduleRole(schedule: ISchedule, targetArn: string): iam.IRole {
+  const stack = Stack.of(schedule);
+  const arn = Token.isUnresolved(targetArn) ? stack.resolve(targetArn).toString() : targetArn;
+  const hash = md5hash(arn).slice(0, 6);
+  const id = 'SchedulerRoleForTarget-' + hash;
+  const existing = stack.node.tryFindChild(id) as iam.IRole;
+
+  const principal = new iam.PrincipalWithConditions(new iam.ServicePrincipal('scheduler.amazonaws.com'), {
+    StringEquals: {
+      'aws:SourceAccount': schedule.env.account,
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:154]: @Jacco I have removed condition for checking sourceArn as it created a circular dependency between schedule -> target -> role -> schedule. 

In your original prototype it was implemented as: 
```
 const scheduleArn = Arn.format({
    service: 'scheduler',
    resourceName: `${schedule.group.groupName}/${schedule.scheduleName}`,
    resource: 'schedule'
  }, Stack.of(schedule));
  const principal = new iam.PrincipalWithConditions(new iam.ServicePrincipal('scheduler.amazonaws.com'), {
    'StringEquals': {
        "aws:SourceArn": scheduleArn,
        "aws:SourceAccount": schedule.env.account,
    }
  })
```

It didn't create a dependency cycle in your code, as now the `scheduleName` is a reference to the CFN resource instead of just a string:
```
    this.scheduleName = this.getResourceNameAttribute(resource.ref);
```

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+  // in the account where the schedule is created.
+  if (sameEnvDimension(schedule.env.account, queue.env.account)) {
+    const policyStatementId = `AllowSchedule${Names.nodeUniqueId(schedule.node)}`;
+
+    queue.addToResourcePolicy(new iam.PolicyStatement({
+      sid: policyStatementId,
+      principals: [new iam.ServicePrincipal('events.amazonaws.com')],
+      effect: iam.Effect.ALLOW,
+      actions: ['sqs:SendMessage'],
+      resources: [queue.queueArn],
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:193]: @Jacco  similarly, there was an option to create a condition for `sourceArn` here. But it would create a dependency cycle between target and schedule.

File: packages/@aws-cdk/aws-scheduler-alpha/test/group.test.ts [hunk 8]
```diff
@@ -15,9 +16,18 @@ describe('Schedule Group', () => {
...
+    role = new iam.Role(stack, 'Role', {
+      roleName: 'someRole',
+      assumedBy: new iam.AccountRootPrincipal(),
+    });
+    func = new lambda.Function(stack, 'MyLambda', {
+      code: new lambda.InlineCode('foo'),
+      handler: 'index.handler',
+      runtime: lambda.Runtime.NODEJS_14_X,
+      tracing: lambda.Tracing.PASS_THROUGH,
+    });
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/test/group.test.ts:30]: Had to specify environment here, otherwise schedule creation failed with an error that target and schedule should be in the same region / account. That's because we consider target and schedule to be in different accounts if one of their accounts is unresolved. 

See `sameEnvDimension` implementation. 

Let me know if that's a problem

File: packages/@aws-cdk/aws-scheduler-alpha/test/input.test.ts [hunk 9]
```diff
@@ -12,18 +14,22 @@ describe('schedule target input', () => {
...
-    role = iam.Role.fromRoleArn(stack, 'Role', 'arn:aws:iam::123456789012:role/johndoe');
-    func = lambda.Function.fromFunctionArn(stack, 'Function', 'arn:aws:lambda:us-east-1:123456789012:function/somefunc');
+    const app = new App();
+    stack = new Stack(app);
+    func = new lambda.Function(stack, 'MyLambda', {
+      code: new lambda.InlineCode('foo'),
+      handler: 'index.handler',
+      runtime: lambda.Runtime.NODEJS_14_X,
+      tracing: lambda.Tracing.PASS_THROUGH,
+    });
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/test/input.test.ts:24]: See https://github.com/aws/aws-cdk/pull/26575/files#r1291166569

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+export function bindBaseTargetConfig(props: ScheduleTargetBaseProps) {
+  let { deadLetterQueue } = props;
+
+  return {
+    deadLetterConfig: deadLetterQueue ? { arn: deadLetterQueue?.queueArn } : undefined,
+    retryPolicy: renderRetryPolicy(props.maxEventAge, props.retryAttempts),
+  };
+}
+
+const roleCache: {[key: string]: iam.Role} = {};
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:136]: why a global prop? Can you identify what the problem is that requires a global prop and if there are any other options? We can discuss decision making in this thread.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+
+/**
+ * Obtain the Role for the EventBridge Scheduler event
+ *
+ * If a role already exists, it will be returned. This ensures that if multiple
+ * events have the same target, they will share a role.
+ * @internal
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:143]: no need for `@internal` since this isn't exported

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+   * The templated target type for the Amazon SQS SendMessage API Operation
+   */
+  readonly sqsParameters?: CfnSchedule.SqsParametersProperty;
+}
+
+/**
+ * Bind props to base schedule target config.
+ * @internal
+ */
+export function bindBaseTargetConfig(props: ScheduleTargetBaseProps) {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:127]: why are all of these exported functions, instead of belonging to a construct?

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+    Annotations.of(schedule).addWarning(`Cannot add a resource policy to your dead letter queue associated with schedule ${schedule.scheduleName} because the queue is in a different account. You must add the resource policy manually to the dead letter queue in account ${queue.env.account}.`);
+  }
+}
+
+/**
+ * Whether two string probably contain the same environment dimension (region or account)
+ *
+ * Used to compare either accounts or regions, and also returns true if both
+ * are unresolved (in which case both are expted to be "current region" or "current account").
+ * @internal
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:205]: same, no `@internal`

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+    }
+    maxAttempts = maximumRetryAttempts;
+  }
+  return {
+    maximumEventAgeInSeconds: maxAge,
+    maximumRetryAttempts: maxAttempts,
+  };
+}
+
+export namespace targets {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:240]: i thought we decided on putting these in `aws-scheduler-targets-alpha`?

File: packages/@aws-cdk/aws-scheduler-alpha/test/input.test.ts [hunk 15]
```diff
@@ -75,12 +80,11 @@ describe('schedule target input', () => {
   test('create an input from object with a ref', () => {
     new Schedule(stack, 'MyScheduleDummy', {
       schedule: expr,
-      target: new targets.LambdaInvoke({
-        role,
+      target: new targets.LambdaInvoke(func, {
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/test/input.test.ts:83]: just confirming these diffs are not breaking changes because `targets.LambdaInvoke` was never exposed.

**Review by kaizencc — COMMENTED** [review]
A few minor comments mostly to get more information. But my main comment is that I would like to see the README first, as that is usually the entrypoint to the PR as a reviewer.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts (hunk 14 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:240]: Yeah, you are right. I just wanted to get the constructs right for this commit to get feedback. Forgot to mention in the review notes. Will put the lambda targets in the separate package and the readme with the next commit.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts (hunk 10 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts:136]: good catch, I do not think we need it. It was used in the prototype code to find the previously created role and update the assumed role policy. We could just get the previously created IAM roles with `stack.node.tryFindChild(id) as iam.Role` and that's it. I will check that in the new commit (and add a unit test).

File: packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts (hunk 12 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/target.ts]: Moved now to abstract class

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 16]
```diff
@@ -1,6 +1,100 @@
...
+   *
+   * @deafult - By default a schedule will be associated with the `default` group.
+   */
+  readonly group?: IGroup;
+}
+
+/**
+ * An EventBridge Schedule
+ */
+export class Schedule extends Resource implements ISchedule {
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:59]: @kaizencc I have moved `Schedule` class out of private zone. Do we need to add a note to readme or anything?

File: packages/@aws-cdk/aws-scheduler-alpha/rosetta/default.ts-fixture [hunk 17]
```diff
@@ -6,8 +6,10 @@ import * as iam from 'aws-cdk-lib/aws-iam';
 import * as kms from 'aws-cdk-lib/aws-kms';
 import * as sqs from 'aws-cdk-lib/aws-sqs';
 import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
+// This import will fail as aws-scheduler-alpha does not depend on aws-scheduler-targets-alpha to avoid cyclic dependency 
+// import * as targets from '@aws-cdk/aws-scheduler-targets-alpha';
```
Comment (filletofish) [inline packages/@aws-cdk/aws-scheduler-alpha/rosetta/default.ts-fixture:10]: Calling this out explicitly in comments too @kaizencc. Do you have any ideas for how to handle that?

Comment (kaizencc) [reply packages/@aws-cdk/aws-scheduler-alpha/rosetta/default.ts-fixture:10]: in package.json we can add 

```json
  "jsiiRosetta": {
    "exampleDependencies": {
      "@aws-cdk/aws-schduler-targets-alpha": "^2.92.0",
    }
  },
```

File: packages/@aws-cdk/aws-scheduler-alpha/README.md [hunk 18]
```diff
@@ -156,7 +161,30 @@ const input = ScheduleTargetInput.fromText(text);
...
 
-TODO: Not yet implemented. See section in [L2 Event Bridge Scheduler RFC](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0474-event-bridge-scheduler-l2.md)
+An execution role is an IAM role that EventBridge Scheduler assumes in order to interact with other AWS services on your behalf.
+
+The classes for templated schedule targets automatically create an IAM role with all the minimum necessary
+permissions to interact with the templated target. If you wish you may specify your own IAM role, then the templated targets
+will grant minimal required permissions. For example: for invoking Lambda function target `LambdaInvoke` will grant
+execution IAM role permission to `lambda:InvokeFunction`.
+
+```text
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/README.md:171]: this should be able to be a ts test once we add

```json
"jsiiRosetta": {
    "exampleDependencies": {
      "@aws-cdk/aws-schduler-targets-alpha": "^2.92.0",
    }
  },
```

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts [hunk 19]
```diff
@@ -1,6 +1,100 @@
...
+  public readonly group?: IGroup;
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:60]: docstrings for these public props

**Review by kaizencc — COMMENTED** [review]
@filletofish I'm ready to approve this, just some minor stuff surrounding docs. 

Next up **has** to be adding integ tests where appropriate. We won't accept any more additions until the integ tests are up and running.

File: packages/@aws-cdk/aws-scheduler-alpha/rosetta/default.ts-fixture (hunk 17 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/rosetta/default.ts-fixture]: Thanks! That worked.

File: packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts (hunk 16 above)
Comment (filletofish) [reply packages/@aws-cdk/aws-scheduler-alpha/lib/schedule.ts:68]: Added now

File: packages/@aws-cdk/aws-scheduler-targets-alpha/test/target.test.ts [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+            ],
+          },
+        ],
+      },
+      Roles: ['someRole'],
+    });
+  });
+
+  test('throws when lambda function is imported from different account', () => {
+    const importedFunc = lambda.Function.fromFunctionArn(stack, 'ImportedFunction', 'arn:aws:lambda:us-east-1:111122223333:function/somefunc');
```
Comment (kaizencc) [inline packages/@aws-cdk/aws-scheduler-targets-alpha/test/target.test.ts:323]: TIL as well, but take a look at this file [here](https://github.com/aws/aws-cdk/blob/apigateway-sagemaker-integration/.gitallowed)

tl;dr: cross env tests need to use `234567890123`

**Review by kaizencc — DISMISSED** [review]

**Comment by kaizencc** [comment]
```
@aws-cdk/aws-lambda-python-alpha: ImportError: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'OpenSSL 1.0.2k-fips  26 Jan 2017'. See: https://github.com/urllib3/urllib3/issues/2168
```

wtf?

**Comment by kaizencc** [comment]
@filletofish still another error related to this PR it seems

**Review by kaizencc — APPROVED** [review]
